### PR TITLE
[7.0.x] service CIDR update

### PR DESCRIPTION
**Pull request by a-palchikov**
## Description

This PR implements changes to fix support for updating service CIDR on a live cluster.

## Type of change

* Regression fix (non-breaking change which fixes a regression)
* This change has a user-facing impact

## Linked tickets and other PRs

* Updates http://github.com/gravitational/gravity/issues/936
* Requires https://github.com/gravitational/gravity.e/pull/4327, https://github.com/gravitational/planet/pull/722.
* Ports https://github.com/gravitational/gravity/pull/1839, https://github.com/gravitational/gravity/pull/1864.

## TODOs

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Write documentation
- [ ] Address review feedback

## Testing done

Install `6.1.33` 3-node cluster.

Original configuration:
```
$ gravity resource get clusterconfiguration
Configuration
=============
Pod IP Range:         10.244.0.0/16
Service IP Range:     10.100.0.0/16
Cloud
-----
Provider:     onprem
```

Update service CIDR to `10.200.0.0/16`:

```yaml
kind: ClusterConfiguration
version: v1
spec:
  global:
    # CIDR range for Services in Cluster
    serviceCIDR: "10.200.0.0/16"
```
```
$ gravity resource create servicecidr.yaml  --confirm
...
$ gravity resource get clusterconfiguration
Configuration
=============
Pod IP Range:         10.244.0.0/16
Service IP Range:     10.200.0.0/16
Cloud
-----
Provider:     onprem
```

Upgrade to the version from this branch and update service CIDR again:

```
kind: ClusterConfiguration
version: v1
spec:
  global:
    # CIDR range for Services in Cluster
    serviceCIDR: "10.203.0.0/16"
```

```
$ gravity resource create servicecidr-203.yaml  --confirm
...
$ gravity resource get clusterconfiguration
Configuration
=============
Pod IP Range:         10.244.0.0/16
Service IP Range:     10.203.0.0/16
Cloud
-----
Provider:     onprem

$ kubectl get svc --all-namespaces
NAMESPACE     NAME                     TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                      AGE
default       kubernetes               ClusterIP      10.203.0.1       <none>        443/TCP                      46m
kube-system   bandwagon                ClusterIP      10.203.87.106    <none>        80/TCP                       46m
kube-system   gravity-site             LoadBalancer   10.203.224.97    <pending>     3009:32009/TCP               46m
kube-system   kube-dns                 ClusterIP      10.203.226.126   <none>        53/UDP,53/TCP                46m
kube-system   kube-dns-worker          ClusterIP      10.203.130.94    <none>        53/UDP,53/TCP                46m
kube-system   kubelet                  ClusterIP      None             <none>        10250/TCP                    22h
kube-system   log-collector            ClusterIP      10.203.165.195   <none>        8083/TCP                     46m
kube-system   lr-aggregator            ClusterIP      10.203.56.8      <none>        9966/TCP                     46m
monitoring    alertmanager-main        ClusterIP      10.203.174.83    <none>        9093/TCP                     46m
monitoring    alertmanager-operated    ClusterIP      None             <none>        9093/TCP,9094/TCP,9094/UDP   22h
monitoring    grafana                  ClusterIP      10.203.142.90    <none>        3000/TCP                     46m
monitoring    kube-state-metrics       ClusterIP      None             <none>        8443/TCP,9443/TCP            22h
monitoring    nethealth                ClusterIP      10.203.197.130   <none>        9801/TCP                     46m
monitoring    node-exporter            ClusterIP      None             <none>        9100/TCP                     22h
monitoring    prometheus-adapter       ClusterIP      10.203.186.214   <none>        443/TCP                      46m
monitoring    prometheus-k8s           ClusterIP      10.203.39.15     <none>        9090/TCP                     46m
monitoring    prometheus-operated      ClusterIP      None             <none>        9090/TCP                     22h
monitoring    prometheus-operator      ClusterIP      None             <none>        8080/TCP                     22h
openebs       admission-server-svc     ClusterIP      10.203.85.114    <none>        443/TCP                      46m
openebs       maya-apiserver-service   ClusterIP      10.203.235.177   <none>        5656/TCP                     46m
```

### CODE REVIEW

File: lib/allocator/allocator.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package allocator manages allocation of resources.
+// Modified from https://github.com/kubernetes/kubernetes/blob/v1.15.12/pkg/registry/core/service/allocator/allocator.go
+package allocator
```
Comment (r0mant) [inline lib/allocator/allocator.go:20]: I wonder if we should place it in lib/utils rather than in just lib.

**Review by r0mant — APPROVED** [review]
Btw, shouldn't we also restart all pods after changing dns services, or otherwise how are they gonna new resolv.conf injected?

File: lib/allocator/allocator.go (hunk 1 above)
Comment (a-palchikov) [reply lib/allocator/allocator.go:20]: `utils` is a collection of disparate implementations with no single purpose. Usually, packages with generic names like this are discouraged in favor of more specific ones. `Allocator` in this context is a generic resource allocator with a specific purpose which I believe has its place at the top of the packages directory.

**Comment by a-palchikov** [comment]
@r0mant not sure I understand the concern. The way it works is the operation will roll out new service CIDR configuration per-node, and when a new node starts, the container agent will match the active service CIDR with a corresponding kube-dns service (depending on the node role), and save the address of the DNS service into the environment file (`/run/dns.env`) which kubelet gets as a prerequisite to start. 
And due to the drain step, the Pods should start off fresh.

**Pull request merged**